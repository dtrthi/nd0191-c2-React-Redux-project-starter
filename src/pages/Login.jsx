import { useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { useLocation, useNavigate } from "react-router";
import { login } from "../actions/authedUser";
import { handleLoadingUsers } from "../actions/users";

const Login = () => {
  const dispatch = useDispatch();
  const authedUser = useSelector((state) => state.authedUser);
  const users = useSelector((state) => state.users);
  const loading = useSelector((state) => state.loading.default);
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
  const [impersonateUser, setImpersonateUser] = useState("");
  const navigate = useNavigate();
  const { search } = useLocation();
  const query = new URLSearchParams(search);
  const redirect = query.get("redirect") || "/";
  useEffect(() => {
    dispatch(handleLoadingUsers());
  }, []);

  useEffect(() => {
    if (authedUser && !loading) {
      navigate(redirect);
    }
  }, [navigate, authedUser, loading, redirect]);

  const handleLogin = async (e) => {
    e.preventDefault();
    let user = null;
    if (users[username] && users[username].password === password) {
      user = users[username];
    }
    if (!user) {
      alert("Username or password incorrect.");
      return;
    }
    dispatch(login(user));
    navigate(redirect);
  };

  const handleImpersonate = () => {
    if (users[impersonateUser]) {
      dispatch(login(users[impersonateUser]));
      navigate(redirect);
    }
  };

  if (authedUser) return null;
  return (
    <div className="d-flex flex-column">
      <h2>Login</h2>
      <form onSubmit={handleLogin}>
        <div>
          <label>
            User
            <br />
            <input
              required
              value={username}
              onChange={(e) => setUsername(e.target.value)}
              data-testid="username"
            />
          </label>
        </div>
        <div>
          <label>
            Password
            <br />
            <input
              type="password"
              required
              value={password}
              onChange={(e) => setPassword(e.target.value)}
              data-testid="password"
            />
          </label>
        </div>
        <button className="btn btn-primary" disabled={loading} type="submit">
          Login
        </button>
      </form>

      <div className="d-flex flex-column m-auto">
        <h2>Impersonate</h2>
        <select
          value={impersonateUser}
          onChange={(e) => setImpersonateUser(e.target.value)}
        >
          <option value="">Select user</option>
          {Object.values(users).map((user) => (
            <option key={user.id} value={user.id}>
              {user.name}
            </option>
          ))}
        </select>
        <button
          className="btn btn-primary"
          disabled={loading}
          onClick={handleImpersonate}
        >
          Login
        </button>
      </div>
    </div>
  );
};

export default Login;
