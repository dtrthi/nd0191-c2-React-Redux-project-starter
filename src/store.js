import { legacy_createStore } from "redux";
import middlewares from "./middlewares";
import reducers from "./reducers";

export default legacy_createStore(reducers, middlewares);
