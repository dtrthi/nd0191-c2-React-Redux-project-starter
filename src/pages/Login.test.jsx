import "@testing-library/jest-dom/vitest";
import { fireEvent, render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import store from "../store";
import Login from "./Login";

describe("Login", () => {
  it("should match snapshot", () => {
    const component = render(
      <Provider store={store}>
        <MemoryRouter>
          <Login />
        </MemoryRouter>
        ,
      </Provider>,
    );
    expect(component).toMatchSnapshot();
  });

  it("should input username and password", () => {
    render(
      <Provider store={store}>
        <MemoryRouter>
          <Login />
        </MemoryRouter>
        ,
      </Provider>,
    );
    const usernameInput = screen.getByTestId("username");
    const passwordInput = screen.getByTestId("password");
    fireEvent.change(usernameInput, { target: { value: "user" } });
    fireEvent.change(passwordInput, { target: { value: "pass" } });
    expect(screen.getByTestId("username").value).toBe("user");
    expect(screen.getByTestId("password").value).toBe("pass");
  });

  it("should have username and password", () => {
    render(
      <Provider store={store}>
        <MemoryRouter>
          <Login />
        </MemoryRouter>
        ,
      </Provider>,
    );
    expect(screen.getByTestId("username")).toBeInTheDocument();
    expect(screen.getByTestId("password")).toBeInTheDocument();
  });
});
