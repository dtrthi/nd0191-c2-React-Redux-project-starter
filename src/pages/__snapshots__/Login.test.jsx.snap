// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Login > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="d-flex flex-column"
      >
        <h2>
          Login
        </h2>
        <form>
          <div>
            <label>
              User
              <br />
              <input
                data-testid="username"
                required=""
                value=""
              />
            </label>
          </div>
          <div>
            <label>
              Password
              <br />
              <input
                data-testid="password"
                required=""
                type="password"
                value=""
              />
            </label>
          </div>
          <button
            class="btn btn-primary"
            disabled=""
            type="submit"
          >
            Login
          </button>
        </form>
        <div
          class="d-flex flex-column m-auto"
        >
          <h2>
            Impersonate
          </h2>
          <select>
            <option
              value=""
            >
              Select user
            </option>
          </select>
          <button
            class="btn btn-primary"
            disabled=""
          >
            Login
          </button>
        </div>
      </div>
      ,
    </div>
  </body>,
  "container": <div>
    <div
      class="d-flex flex-column"
    >
      <h2>
        Login
      </h2>
      <form>
        <div>
          <label>
            User
            <br />
            <input
              data-testid="username"
              required=""
              value=""
            />
          </label>
        </div>
        <div>
          <label>
            Password
            <br />
            <input
              data-testid="password"
              required=""
              type="password"
              value=""
            />
          </label>
        </div>
        <button
          class="btn btn-primary"
          disabled=""
          type="submit"
        >
          Login
        </button>
      </form>
      <div
        class="d-flex flex-column m-auto"
      >
        <h2>
          Impersonate
        </h2>
        <select>
          <option
            value=""
          >
            Select user
          </option>
        </select>
        <button
          class="btn btn-primary"
          disabled=""
        >
          Login
        </button>
      </div>
    </div>
    ,
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
